File: Middleware/Thrift-homework/src/main/py/memory_test.py
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol
from thrift.protocol import TCompactProtocol
from thrift.protocol import TJSONProtocol
from thrift_client import DataGenerator
import sys


def print_error():
    """A function printing an error when bad arguments provided"""

    print("Bad arguments provided.")
    print("The program needs a serialisation method to be provided as program argument: binary | json | compact")


def read_protocol(protocol_type, trans):
    """
    A function returning the protocol based on an argument provided.

    :param protocol_type: the protocol type
    :param trans: the memory buffer
    :return: the protocol of the required type
    """

    if protocol_type == "binary":
        return TBinaryProtocol.TBinaryProtocol(trans)
    elif protocol_type == "json":
        return TJSONProtocol.TJSONProtocol(trans)
    elif protocol_type == "compact":
        return TCompactProtocol.TCompactProtocol(trans)
    return None


if __name__ == '__main__':
    """
    A simple script running memory test, that is counting how many bytes are written to TMemoryBuffer
    using different serialisation methods while sending 1000000 elements.
    """

    if len(sys.argv) < 2:
        print_error()
        exit(1)

    trans = TTransport.TMemoryBuffer()

    protocol_type = sys.argv[1]
    proto = read_protocol(protocol_type, trans)
    if proto is None:
        print_error()
        exit(1)

    data = DataGenerator().generate_list(1000000)

    for x in data:
        proto.writeI32(x)

    print(f"Using {protocol_type} and sending 1000000 elements wrote {trans.cstringio_buf.tell()} "
          f"bytes to the TMemoryBuffer")

File: Middleware/Thrift-homework/src/main/java/xenoteo/com/github/MemoryTest.java
package xenoteo.com.github;

import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TCompactProtocol;
import org.apache.thrift.protocol.TJSONProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TMemoryBuffer;
import xenoteo.com.github.client.DataGenerator;

import java.util.List;

/**
 * A simple class running memory test, that is counting how many bytes are written to TMemoryBuffer
 * using different serialisation methods while sending 1000000 elements.
 */
public class MemoryTest {

    public static void main(String[] args) {
        if (args.length < 1) {
            printError();
            System.exit(1);
        }
        try {
            String protocolType = args[0];

            TMemoryBuffer trans = new TMemoryBuffer(4096);

            TProtocol protocol = readProtocol(protocolType, trans);
            if (protocol == null) {
                printError();
                System.exit(1);
            }

            DataGenerator generator = new DataGenerator();
            List<Integer> list = generator.generateList(1000000);
            for (int x : list){
                protocol.writeI32(x);
            }

            System.out.printf("Using %s protocol and sending 1000000 elements wrote %d bytes to the TMemoryBuffer\n",
                    protocolType, trans.length());
        } catch (TException e) {
            e.printStackTrace();
        }
    }

    /**
     * Prints an error when bad arguments provided.
     */
    private static void printError() {
        System.err.println("Bad arguments provided.");
        System.err.println("The program needs a serialisation method to be provided as program argument: " +
                "binary | json | compact");
    }

    /**
     * Returns the TProtocol based on an argument provided.
     *
     * @param protocol  the string representing a protocol type
     * @param trans  the memory buffer
     * @return the TProtocol of the required type
     */
    private static TProtocol readProtocol(String protocol, TMemoryBuffer trans){
        return switch (protocol) {
            case "binary" -> new TBinaryProtocol(trans);
            case "json" -> new TJSONProtocol(trans);
            case "compact" -> new TCompactProtocol(trans);
            default -> null;
        };
    }
}
